src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,414 @@
+  };
+
+  /**
+   * Implementation of action of Frozen Lake.
+   */
+  class Action
+  {
+   public:
+    enum actions
+    {
+      Up, Down, Left, Right
+    };
+    // To store the action.
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 4;
+  };
+
+  //! Constructor.
+  FrozenLake(size_t const& maxSteps=200,
+              size_t const& nRows=4,
+              size_t const& nCols=4,
+              double const& platformRate=0.8) : 
+    maxSteps(maxSteps),
+    nRows(nRows),
+    nCols(nCols)
+  { 
+    // Preprocess platform rate
+    this->platformRate = std::min(platformRate, 1.0);
+    this->platformRate = std::max(platformRate, 0.0);
+  }
+
+  //  Step function
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    size_t newRow = state.CurRow();
+    size_t newCol = state.CurCol();
+
+    // Copy the state to the next state.
+    nextState = state;
+    
+    // Calculate new row and col.
+    if (action.action == Action::actions::Left)
+      newCol = std::max(nextState.CurCol() - 1, (unsigned long) 0);
+    else if (action.action == Action::actions::Down)
+      newRow = std::min(nextState.CurRow() + 1, nRows - 1);
+    else if (action.action == Action::actions::Right)
+      newCol = std::min(nextState.CurCol() + 1, nCols - 1);
+    else if (action.action == Action::actions::Up)
+      newRow = std::max(nextState.CurRow() - 1, (unsigned long) 0);
+    
+    // Update states by setting new current row and column, 
+    //    and manually set the description of board.
+    nextState.CurRow() = newRow;
+    nextState.CurCol() = newCol;
+
+    // Check if the episode has terminated.
+    bool done = IsTerminal(nextState);
+
+    // Reward agent if it reached the goal.
+    if (done && maxSteps != 0 && newRow == nRows - 1 && 
+        newCol == nCols - 1)
+      return 1.0;
+    else if (done && boardDescription[newRow][newCol] == 'H')
+      return -1.0;
+
+    // Reward 0 otherwise.
+    return 0.0;
+  }
+
+  /**
+   * Dynamics of Frozen Lake. Get reward based on current 
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state, const Action& action)
+  {
+    State nextState;
+    return Sample(state, action, nextState);
+  }
+
+  /**
+   * Initial state representation is the current position at (0, 0) 
+   * and a new random bord is created.
+   *
+   * @return Initial state for each episode.
+   */
+  State InitialSample()
+  {
+    stepsPerformed = 0;
+    boardDescription = generateRandomBoard();
+    // Log::Info << boardDescription;
+    return State(nRows, nCols);
+  }
+
+  /**
+   * Initial state representation is the current position at (0, 0) 
+   * and a new random bord is created.
```

Comment:
Pedantic style issue, can you add an extra empty line between the method and parameter description.